### PR TITLE
[dbnode] Add index regexp DFA compilation cache to avoid allocating DFAs for same expressions

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

This reduces allocations building regexp DFA for the same regexp queries one each index query, instead using an LRU to keep the most frequently queried regexp DFA's around for reuse. Also includes metrics.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/m3ninx/index/regexp.go [hunk 1]
```diff
@@ -21,17 +21,23 @@
 package index
 
 import (
+	"context"
 	"fmt"
 	re "regexp"
 	"regexp/syntax"
+	"sync"
+
+	"github.com/uber-go/tally"
```
Comment (arnikola) [inline src/m3ninx/index/regexp.go:30]: nit; 3rd group

File: src/x/cache/nop_cache.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// THE SOFTWARE.
+
+package cache
+
+import (
+	"context"
+	"time"
+)
+
+// NewNop returns a new nop cache.
```
Comment (arnikola) [inline src/x/cache/nop_cache.go:28]: nit: `Noop`

File: src/x/cache/lru_cache.go [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+	}
+
+	return &LRU{
+		ttl:               ttl,
+		now:               now,
+		maxEntries:        maxEntries,
+		cacheErrors:       opts.CacheErrorsByDefault,
+		concurrencyLeases: concurrencyLeases,
+		metrics: &lruCacheMetrics{
+			entries:       tallyScope.Gauge(entriesGauge),
```
Comment (arnikola) [inline src/x/cache/lru_cache.go:201]: nit: may be better to have `tallyScope` be scoped to `cache` rather than requiring one that's scoped to be passed in

Comment (robskillington) [reply src/x/cache/lru_cache.go:203]: Yeah true, done 👍

**Review by rallen090 — COMMENTED** [review]
LGTM

File: src/x/cache/lru_cache.go [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+) (interface{}, error) {
+	// If we're limiting overall concurrency, acquire a concurrency lease
+	if c.concurrencyLeases != nil {
+		select {
+		case <-ctx.Done():
+			return c.cacheLoadComplete(key, time.Time{}, nil, UncachedError{ctx.Err()})
+		case <-c.concurrencyLeases:
+		}
+
+		defer func() { c.concurrencyLeases <- struct{}{} }()
```
Comment (arnikola) [inline src/x/cache/lru_cache.go:472]: nit; maybe slightly safer as:

```suggestion
		select {
		case <-ctx.Done():
			return c.cacheLoadComplete(key, time.Time{}, nil, UncachedError{ctx.Err()})
		case <-c.concurrencyLeases:
			defer func() { c.concurrencyLeases <- struct{}{} }()
		}

```

File: src/x/cache/lru_cache.go [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+}
+
+// As returns true if the caller is asking for the error as an uncached error
+func (e UncachedError) As(target interface{}) bool {
+	if uncached, ok := target.(*UncachedError); ok {
+		uncached.Err = e.Err
+		return true
+	}
+	return false
+}
```
Comment (arnikola) [inline src/x/cache/lru_cache.go:83]: nit; dont think we use these

File: src/x/cache/lru_cache.go [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+
+// As returns true if the caller is asking for the error as an cached error
+func (e CachedError) As(target interface{}) bool {
+	if cached, ok := target.(*CachedError); ok {
+		cached.Err = e.Err
+		return true
+	}
+	return false
+}
```
Comment (arnikola) [inline src/x/cache/lru_cache.go:109]: nit; dont think we use these

File: src/x/cache/lru_cache.go [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+
+	entry, exists := c.entries[key]
+
+	// If a load is already in progress, tell the caller to wait for it to finish.
+	if exists && entry.loadingCh != nil {
+		return nil, false, entry.loadingCh, nil
+	}
+
+	// If the entry exists and has not expired, it's a hit - return it to the caller
+	if exists && entry.expiresAt.After(c.now()) {
```
Comment (arnikola) [inline src/x/cache/lru_cache.go:319]: Does this need to check that it's not expired? You'll likely be regenerating the same entry anyway here

File: src/x/cache/lru_cache.go [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+		if load {
+			if loader == nil {
+				return nil, ErrEntryNotFound
+			}
+
+			return c.tryLoad(ctx, key, loader)
+		}
+
+		// There is an entry and it's valid, return it.
+		return value, nil
```
Comment (arnikola) [inline src/x/cache/lru_cache.go:283]: Do we want this to update the entry TTL?

**Review by arnikola — APPROVED** [review]
LGTM

File: src/x/cache/nop_cache.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
```
Comment (arnikola) [inline src/x/cache/nop_cache.go:1]: nit; update filename too

**Pull request merged**